### PR TITLE
COSBETA-255: Add all trigger questions

### PR DESCRIPTION
**Pull request by TiagoSusanoPinto**
COSBETA-254: Create trigger questions templates
COSBETA-255: Add all trigger questions

## Checklist:

- [ ] Automated checks are passing locally.
- [ ] CHANGELOG updated if change is worth telling users about.
### General testing
- [ ] Test without javascript
- [ ] Test on small screen
### Accessibility testing
- [ ] Works keyboard only
- [ ] Tested with one screen reader
- [ ] Zoom page to 400% - content still visible
- [ ] Disable css - does content make sense and appear in a logical order?
- [ ] Passes automated checker (automated in build or manual)

### CODE REVIEW

File: cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss [hunk 1]
```diff
@@ -0,0 +1,26 @@
+.input-wrapper {
+  display: flex;
```
Comment (jasiekmiko) [inline cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss]: What browsers does cosmetics need to support? If any of the older stuff, flex might not work, in particular, [IE8 is doesn't do it](https://caniuse.com/#search=flex)

File: cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+$govuk-light-grey: #f3f2f1; // New light grey colour
```
Comment (jasiekmiko) [inline cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss]: Is it a `govuk` colour? If not, use own prefix would be preferable, have you guys got one established? `cosmetics-light-grey`?
Same goes for the names of classes defined below

File: cosmetics-web/app/views/trigger_questions/_substance_check.html.slim [hunk 3]
```diff
@@ -0,0 +1,16 @@
+ruby:
+  component_name = @component.notification.is_multicomponent? ? @component.name : "the cosmetic product"
+  default_question = "Does #{component_name} contain #{substance.pluralize}?"
+  default_yes_option = "Yes, it has #{substance.pluralize}"
+
+= form_with model: @question, url: wizard_path, method: :put do |form|
+  .govuk-grid-row
+    .govuk-grid-column-two-thirds
+      = render "form_components/govuk_error_summary", form: form
+      h1.govuk-heading-xl
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/_substance_check.html.slim]: Lets [make this the legend](https://design-system.service.gov.uk/get-started/labels-legends-headings/)

File: cosmetics-web/app/views/trigger_questions/_substance_check_with_condition.html.slim [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  component_name = @component.notification.is_multicomponent? ? @component.name : "the cosmetic product"
+  default_question = "Does #{component_name} contain #{substance.pluralize}?"
+  default_yes_option = "Yes, the product contains #{substance}"
+  default_label_text = "Total amount of #{substance} in the product"
+
+= form_with model: @question, url: wizard_path, method: :put do |form|
+  .govuk-grid-row
+    .govuk-grid-column-two-thirds
+      = render "form_components/govuk_error_summary", form: form
+      h1.govuk-heading-xl
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/_substance_check_with_condition.html.slim]: Lets [make this the legend](https://design-system.service.gov.uk/get-started/labels-legends-headings/)

File: cosmetics-web/app/views/trigger_questions/_substance_list.html.slim [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      - if @errors.present?
+        = render "components/govuk_error_summary", titleText: "There is a problem", errorList: @errors
+      h1.govuk-heading-xl
+        = local_assigns[:question].present? ? question : "List the #{substance.pluralize} and their concentration"
+      = form.fields_for :trigger_question_elements do |answer_form|
+        = answer_form.hidden_field :answer_order, value: answer_form.object.answer_order
+        = answer_form.hidden_field :element_order, value: answer_form.object.element_order
+        = answer_form.hidden_field :element, value: answer_form.object.element
+        - if answer_form.object.element == "inciname"
+          .govuk-grid-column-two-thirds
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/_substance_list.html.slim]: You shouldn't have a `.govuk-grid-column-two-thirds` inside a `.govuk-grid-column-two-thirds` - unless it's also got a `.govuk-grid-row`!
I.e., bad:
```
.govuk-grid-row
  .govuk-grid-column-two-thirds
    .govuk-grid-column-two-thirds
```
good:
```
.govuk-grid-row
  .govuk-grid-column-two-thirds
    .govuk-grid-row
      .govuk-grid-column-two-thirds
```
or 
```
.govuk-grid-row
  .govuk-grid-column-two-thirds
  .govuk-grid-column-two-thirds
```

File: cosmetics-web/app/views/trigger_questions/_substance_list.html.slim [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+            = render "form_components/govuk_input", form: answer_form, key: :answer,
+                    label: { text: substance.upcase_first }
+        - else
+          .govuk-grid-column-one-third
+            - errors = answer_form.object.errors
+            .govuk-form-group class=("govuk-form-group--error" if errors.present?)
+              = answer_form.label :answer, "Concentration", class: "govuk-label"
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/_substance_list.html.slim]: There's a `components/govuk_label` which you could use here. it sets a lot of the boilerplate you've got, including the errors stuff

File: cosmetics-web/app/views/trigger_questions/_substance_list.html.slim [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+              - if errors.present?
+                span.govuk-error-message id="trigger_question_trigger_question_elements.answer-error"
+                  = errors.full_messages_for(:answer).first
+              .input-wrapper
+                = answer_form.text_field :answer, class: "govuk-input govuk-c-input"
+                span.govuk-input.govuk-c-input__suffix
+                  | % w/w
+  .govuk-form-group
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/_substance_list.html.slim]: Correct me if I'm wrong, but I don't think this div adds anything here? :)

File: cosmetics-web/app/views/trigger_questions/_substance_text_input.html.slim [hunk 8]
```diff
@@ -0,0 +1,15 @@
+- heading = capture do
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/_substance_text_input.html.slim]: Do you need the capture here, or can you just get away with passing text and classes to the component?

File: cosmetics-web/app/views/trigger_questions/select_ph_range.html.slim [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+- component_name = @component.notification.is_multicomponent? ? @component.name : "the cosmetic product"
+= form_with url: wizard_path, method: :put do |form|
+  .govuk-grid-row
+    .govuk-grid-column-two-thirds
+      - if @errors.present?
+        = render "components/govuk_error_summary", titleText: "There is a problem", errorList: @errors
+      h1.govuk-heading-xl
+        | What is the pH of
+        =< component_name
+        | ?
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/select_ph_range.html.slim]: Would this be easier to write as `h1.govuk-heading-xl = "What is the pH of #{component_name}?"`?

File: cosmetics-web/app/views/trigger_questions/select_ph_range.html.slim [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+- content_for :after_header do
+  = link_to "Back", previous_wizard_path, class: "govuk-back-link"
+
+- component_name = @component.notification.is_multicomponent? ? @component.name : "the cosmetic product"
+= form_with url: wizard_path, method: :put do |form|
+  .govuk-grid-row
+    .govuk-grid-column-two-thirds
+      - if @errors.present?
+        = render "components/govuk_error_summary", titleText: "There is a problem", errorList: @errors
+      h1.govuk-heading-xl
```
Comment (jasiekmiko) [inline cosmetics-web/app/views/trigger_questions/select_ph_range.html.slim]: Should be the fieldset of the radios

File: cosmetics-web/config/webpacker.yml [hunk 11]
```diff
@@ -69,7 +69,7 @@ development:
     headers:
       'Access-Control-Allow-Origin': '*'
     watch_options:
-      poll: true
+#      poll: true
```
Comment (jasiekmiko) [inline cosmetics-web/config/webpacker.yml]: Is this intentional? I believe this breaks hot-reload in dev on windows..!

Comment (TiagoSusanoPinto) [reply cosmetics-web/config/webpacker.yml]: Ups, I forgot to revert it...

File: cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss (hunk 1 above)
Comment (TiagoSusanoPinto) [reply cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss]: According to Nasir, we don't need to support IE8, but as long it's not a functional problem, it is OK.

But to be honest, I'm not sure what's the solution to show a suffix/prefix in the input field that will work in all environments.

Comment (jasiekmiko) [reply cosmetics-web/app/assets/application/stylesheets/input-prefix-suffix.scss]: If IE8 isn't a problem, then I'm all for using `flex`, cause it's great - precisely for that reason! :)

For the sake of academics.. if IE8 was required, then *a* fallback is using table. It's onbviously not ideal in terms of semantics, but it is an option which I think would definitely work. First I'd investigate some combination of wrapper divs, floats, before/after pseudo-elements, etc. Finally, you can hope to progressively enhance either of these solutions to flex using [one of these techniques](https://medium.com/css-mine/flexbox-how-to-deal-with-older-browsers-fbf6eb8c7a65).

But yes, in this case, I think we just want to use flex directly :)

File: cosmetics-web/app/views/trigger_questions/add_alkaline_agents.html.slim [hunk 12]
```diff
@@ -0,0 +1,6 @@
+- content_for :page_title, "Add anti-dandruff agents"
```
Comment (tom-bell-softwire) [inline cosmetics-web/app/views/trigger_questions/add_alkaline_agents.html.slim]: Should this be "anti-dandruff" ?

File: cosmetics-web/config/initializers/inflections.rb [hunk 13]
```diff
@@ -12,4 +12,6 @@
 
 ActiveSupport::Inflector.inflections(:en) do |inflect|
   inflect.irregular 'person', 'persons'
+  inflect.irregular 'compound releasing hydrogen peroxide', 'compounds releasing hydrogen peroxide'
+  inflect.irregular 'essential oil, camphor, menthol, or eucalyptol', 'essential oils, camphor, menthol, or eucalyptol'
```
Comment (tom-bell-softwire) [inline cosmetics-web/config/initializers/inflections.rb]: Nice!

File: cosmetics-web/spec/controllers/trigger_questions_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,197 @@
+require 'rails_helper'
+
+RSpec.describe TriggerQuestionsController, type: :controller do
+  let(:responsible_person) { create(:responsible_person) }
+  let(:notification) { create(:notification, components: [create(:component)], responsible_person: responsible_person) }
+  let(:multi_component_notification) { create(:notification, components: [create(:component), create(:component)], responsible_person: responsible_person) }
+  let(:component) { notification.components.first }
+  let(:multi_component) { multi_component_notification.components.first }
```
Comment (tom-bell-softwire) [inline cosmetics-web/spec/controllers/trigger_questions_controller_spec.rb]: This is only used once in `:multi_component_params` so I'd be inclined to just inline this code there instead.

File: cosmetics-web/spec/controllers/trigger_questions_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+      let(:valid_answers_with_unpaired_values) {
+        [
+            { answer: "agent 1", answer_order: 0, element_order: 0, element: "inciname", id: answers.first.id },
+            { answer: "5", answer_order: 0, element_order: 0, element: "incivalue", id: answers.second.id },
+            { answer: "agent 2", answer_order: 1, element_order: 0, element: "inciname", id: answers.third.id },
+            { answer: "", answer_order: 1, element_order: 1, element: "incivalue", id: answers.fourth.id }
+        ]
+      }
+
+      let(:unvalid_answers) { [{ answer: "", answer_order: 0, element_order: 0, element: "inciname", id: answers.first.id }] }
```
Comment (tom-bell-softwire) [inline cosmetics-web/spec/controllers/trigger_questions_controller_spec.rb]: Minor typo: "unvalid" -> "invalid"

Also, it'd be good to split this over multiple lines, as you've done for the other examples above.

File: cosmetics-web/spec/controllers/trigger_questions_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,197 @@
+require 'rails_helper'
+
+RSpec.describe TriggerQuestionsController, type: :controller do
```
Comment (tom-bell-softwire) [inline cosmetics-web/spec/controllers/trigger_questions_controller_spec.rb]: These tests are really great!

File: cosmetics-web/app/controllers/trigger_questions_controller.rb [hunk 17]
```diff
@@ -37,6 +96,58 @@ def finish_wizard_path
     end
   end
 
+  def previous_wizard_path
+    if step == steps.first
+      return responsible_person_notification_component_build_path(@component.notification.responsible_person, @component.notification, @component, :select_frame_formulation)
+    end
+
+    previous_step = get_previous_step
+    if !previous_step.nil?
```
Comment (tom-bell-softwire) [inline cosmetics-web/app/controllers/trigger_questions_controller.rb]: You can do `if previous_step.present?` here, which is slightly more readable.

File: cosmetics-web/app/views/trigger_questions/add_alkaline_agents.html.slim (hunk 12 above)
Comment (TiagoSusanoPinto) [reply cosmetics-web/app/views/trigger_questions/add_alkaline_agents.html.slim]: Good catch!

File: cosmetics-web/app/models/trigger_question_element.rb [hunk 18]
```diff
@@ -1,33 +1,41 @@
 class TriggerQuestionElement < ApplicationRecord
   belongs_to :trigger_question
 
-  validates :answer, numericality: true, if: -> { applicable_question? && is_ph? }
-  validates :answer, presence: true, if: -> { applicable_question? && is_ethanol_or_isopropanol? }
-  validate :no_empty_answers_with_the_same_answer_order, if: -> { applicable_question? && :is_inciname_or_incivalue? }
+  validates :answer, numericality: true, if: -> { question_is_applicable? && self.ph? }
+  validates :answer, presence: true, if: -> { question_is_applicable? && answer_is_single_value? }
+  validate :no_empty_answers_within_grouped_answers?, if: -> { question_is_applicable? && :answer_is_grouped_value? }
```
[the remark this answers is not in the document]
Comment (TiagoSusanoPinto) [reply cosmetics-web/app/models/trigger_question_element.rb]: ohhh, my bad! I though that I had changed it, but then with renaming methods, I might had revert it...

**Comment by TiagoSusanoPinto** [comment]
I've just rebase it on top of master, fix merge conflicts, and squash all markups commits into a single one

**Pull request merged**